### ISSUE TITLE
Defining DOCTEST_CONFIG_DISABLE globally causes linking errors

### ISSUE THREAD

**Issue by Niketin** [doctest/doctest#873]
## Description

I created a project based on the executable_dll_and_plugin example where I wanted to try doctest and see how it works when tests are implemented alongside library source code. Everything worked fine with my modifications to the example, until I wanted to try defining the DOCTEST_CONFIG_DISABLE globally (for production use case). I got the following linker error.
```
test_runner.obj : error LNK2019: unresolved external symbol "public: virtual __cdecl doctest::IReporter::~IReporter(void)" (??1IReporter@doctest@@UEAA@XZ) referenced in function "public: virtual void * __cdecl doctest::IReporter::`vector deleting destructor'(unsigned int)" (??_EIReporter@doctest@@UEAAPEAXI@Z) [C:\Users\Niketin\repos\doctest-example\build\dll\test_runner.vcxproj]
C:\Users\Niketin\repos\doctest-example\build\dll\Debug\test_runner.dll : fatal error LNK1120: 1 unresolved externals [C:\Users\Niketin\repos\doctest-example\build\dll\test_runner.vcxproj]
```
I wasn't able to solve this problem. I assume it is either a user error, or something is wrong in doctest, so I decided to open an issue here.

### Steps to reproduce

1. Download and extract [doctest-example.zip](https://github.com/user-attachments/files/16970025/doctest-example.zip)
2. Build it with CMake:
   a. `cmake -B build .`
   b. `cmake --build build`
3. Build should complete fine. Listing tests show all 4 tests.
```
PS C:\Users\Niketin\repos\doctest_example> .\build\test\Debug\executable_dll_and_plugin.exe -ltc
[doctest] listing all test case names
===============================================================================
dll
dll utils
test_runner
executable
===============================================================================
[doctest] unskipped test cases passing the current filters: 4
```
4. Now, add `#define DOCTEST_CONFIG_DISABLE` at the beginning of each source file.
5. Build again. The error should come up.

### Extra information

* doctest version: **v2.4.11**
* Operating System: **Microsoft Windows 10 Pro Version 10.0.19045 Build 19045**
* Compiler+version: **MSVC 19.39.33523.0**

**Comment by Niketin** [comment]
> I can not reproduce. I'm on Ubuntu 23.10, GCC 13.2.0. Note that the repro instructions are not correct, the source files in the zip all define `DOCTEST_CONFIG_DISABLE` and you need to remove those to repro step 3. When I run the executable built from the source code downloaded as is I get no output at all.
> 
> If you can reduce the bug into a minimal example that would be helpful.

You are right @cdeln . I made a mistake in the instructions. The zip file contains the example code in the state where the `DOCTEST_CONFIG_DISABLE` is already defined for every source file. I must have followed the instructions and uploaded the package in the end state. I can give you new, corrected steps that work for me on the above mentioned Windows system.

1. Download and extract [doctest-example.zip](https://github.com/user-attachments/files/16970025/doctest-example.zip)
2. Build it with CMake:
  a. cmake -B build .
  b. cmake --build build
3. The build should fail with the error.
3. Now, remove `#define DOCTEST_CONFIG_DISABLE` at the beginning of each source file.
    Build again. The error should no longer come up.

**Comment by Niketin** [comment]
I will try to minimize the example as soon as possible.

**Comment by cdeln** [comment]
Thank you @Niketin ! :)

**Comment by Niketin** [comment]
Ok, I think this will be something that you can work with more easily as it's minimal patch to the existing example in the doctest code base. If you define the `DOCTEST_CONFIG_DISABLE` for the example `executable_dll_and_plugin`, it no longer builds.
```diff
diff --git a/examples/executable_dll_and_plugin/CMakeLists.txt b/examples/executable_dll_and_plugin/CMakeLists.txt
index 31b5ed2..2944ee2 100644
--- a/examples/executable_dll_and_plugin/CMakeLists.txt
+++ b/examples/executable_dll_and_plugin/CMakeLists.txt
@@ -1,3 +1,5 @@
+add_compile_definitions(DOCTEST_CONFIG_DISABLE)
+
 add_library(implementation SHARED implementation.cpp implementation_2.cpp)
 target_link_libraries(implementation PUBLIC doctest)

```
The error message is the same as before.

I tried it on both compilers from Visual Studio 16 2019 and Visual Studio 17 2022, and also tried Debug and Release configurations separately. Always the same error.

**Comment by cdeln** [comment]
Thanks, this is a perfect MRE. I applied the diff and can still not repro on my setup, but it seems it indeed fails on similar setups, and in particular Windows.

**Comment by jorgenpt** [comment]
I also encountered this, and if anyone else encounters it, you can add the following snippet to make it link correctly:
```cpp
// Work around linker error due to undefined `IReporter::~IReporter` when
// doctest is disabled in DLL configurations.
// https://github.com/doctest/doctest/issues/873
#ifdef DOCTEST_CONFIG_DISABLE
namespace doctest {
DOCTEST_DEFINE_INTERFACE(IReporter)
}
#endif
```
